src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -7215,98 +7213,90 @@ bool SketchObject::insertBSplineKnot(int GeoId, double param, int multiplicity)
                                   "Knot cannot be inserted outside the B-spline parameter range."));
     }
 
-    std::unique_ptr<Part::GeomBSplineCurve> bspline;
-
-    // run the command
+    // Clone the B-spline and insert the knot
+    std::unique_ptr<Part::GeomBSplineCurve> bspline{};
     try {
         bspline.reset(static_cast<Part::GeomBSplineCurve*>(bsp->clone()));
-
         bspline->insertKnot(param, multiplicity);
     }
     catch (const Base::Exception& e) {
         Base::Console().Error("%s\n", e.what());
         return false;
     }
 
-    // once command is run update the internal geometries
-    std::vector<int> delGeoId;
-
-    std::vector<Base::Vector3d> poles = bsp->getPoles();
-    std::vector<Base::Vector3d> newPoles = bspline->getPoles();
-    std::vector<int> poleIndexInNew(poles.size(), -1);
-
-    for (size_t j = 0; j < poles.size(); j++) {
-        const auto it = std::find(newPoles.begin(), newPoles.end(), poles[j]);
-        poleIndexInNew[j] = it - newPoles.begin();
-    }
-    std::replace(poleIndexInNew.begin(), poleIndexInNew.end(), int(newPoles.size()), -1);
-
-    std::vector<double> knots = bsp->getKnots();
-    std::vector<double> newKnots = bspline->getKnots();
-    std::vector<int> knotIndexInNew(knots.size(), -1);
+    // Update the internal geometries, map poles and knots
+    auto findIndex = [](const auto& container, const auto& value) -> std::optional<int> {
+        if (auto it = std::find(container.begin(), container.end(), value); it != container.end()) {
+            return std::distance(container.begin(), it);
+        }
+        return std::nullopt;
+    };
 
-    for (size_t j = 0; j < knots.size(); j++) {
-        const auto it = std::find(newKnots.begin(), newKnots.end(), knots[j]);
-        knotIndexInNew[j] = it - newKnots.begin();
-    }
-    std::replace(knotIndexInNew.begin(), knotIndexInNew.end(), int(newKnots.size()), -1);
+    const auto& poles{bsp->getPoles()};
+    const auto& newPoles{bspline->getPoles()};
+    std::vector<std::optional<int>> poleIndexInNew{std::size(poles)};
+    std::transform(poles.begin(), poles.end(), poleIndexInNew.begin(),
+                   [&newPoles, &findIndex](const auto& pole) { return findIndex(newPoles, pole); });

```

Comment:
There is a chance of slightly more cleanliness if some kind of [currying](https://en.wikipedia.org/wiki/Currying) is possible.
```
std::transform(poles.begin(), poles.end(), poleIndexInNew.begin(), indexFinderFor(newPoles));
```

Where `indexFinderFor(x)(y)` does the same thing as `findIndex(x, y)`.

Can you think of something similar in C++?

Reply:
The `sourceContainer` doesn't seem necessary. It's not used in the lambdas anyway.

I think it's a good trade-off to avoid having to write more lambdas, that too within `std::transform`. I have grown to think they should have cognitive complexity of their own.

Reply:
I don't agree with you as I think it requires a higher cognitive load by having to read a lambda before the context is given. By reading the lambda in place I have  the context I need around it.
The lambda in lambda doesn't make it much easier either.

The first times I encountered std::transform it was a bit much to wrap my brain around since it was new, but now I recognize it as a common token and can see "through it" and I personally think it's quite neat to have the lambda in place in transform, find, etc.

Reply:
> I don't agree with you as I think it requires a higher cognitive load by having to read a lambda before the context is given. By reading the lambda in place I have the context I need around it. The lambda in lambda doesn't make it much easier either.

We indeed don't agree here :). Though I do agree that having the lambda after the context would be a good idea.

Reply:
Then let's agree to disagree 😅

I wonder what would be the best way to continue. Would it be ok that you just pick what you think would be good and ignore the rest and then I'll close this PR?

I only created it because you asked me to show how I would write it and then create the pr. Your previous version is fine. The only big issue I had was that it had a bug, but you've fixed that already 😊

Reply:
> I wonder what would be the best way to continue. Would it be ok that you just pick what you think would be good and ignore the rest and then I'll close this PR?

I would really like to use the newer functionalities where possible, and to that extent my intention was to add your commit over the existing PR. However, at this moment there's a segfault causing bug in this commit, and you'd be the best person to solve it. If you don't have the time, best to close this PR, and then if I figure it out I can always reopen and merge or do something similar.

Reply:
I'll see what I can do and ping you afterwards